rapids_logger/logger.hpp.in

```diff
@@ -0,0 +1,575 @@
+  }
+
+  /**
+   * @brief Log a message at the DEBUG level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void debug(std::string const& format, Args&&... args)
+  {
+    log(level_enum::debug, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the INFO level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void info(std::string const& format, Args&&... args)
+  {
+    log(level_enum::info, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the WARN level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void warn(std::string const& format, Args&&... args)
+  {
+    log(level_enum::warn, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the ERROR level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void error(std::string const& format, Args&&... args)
+  {
+    log(level_enum::error, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the CRITICAL level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void critical(std::string const& format, Args&&... args)
+  {
+    log(level_enum::critical, format, std::forward<Args>(args)...);
+  }
+
+#ifndef @_RAPIDS_LOGGER_MACRO_PREFIX@_BACKWARDS_COMPATIBILITY
+ private:
+#endif
+#ifdef @_RAPIDS_LOGGER_MACRO_PREFIX@_SUPPORTS_LOGGING
+  std::unique_ptr<spdlog::logger> impl;
+#endif
+
+#ifdef @_RAPIDS_LOGGER_MACRO_PREFIX@_BACKWARDS_COMPATIBILITY
+ private:
+#endif
+  // TODO: When we migrate to C++20 we can use std::format and format strings
+  // instead of the printf-style printing used here.
+  /**
+   * @brief Format and log a message.
+   *
+   * This function performs printf-style formatting to avoid the need for fmt
+   * or spdlog's own templated APIs (which would require exposing spdlog
+   * symbols publicly) and then logs the formatted message.
+   *
+   * @param lvl The log level
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void formatted_log(level_enum lvl, std::string const& format, Args&&... args) {
+    auto convert_to_c_string = [](auto&& arg) -> decltype(auto) {
+      using ArgType = std::decay_t<decltype(arg)>;
+      if constexpr (std::is_same_v<ArgType, std::string>) {
+        return arg.c_str();
+      } else {
+        return std::forward<decltype(arg)>(arg);
+      }
+    };
+
+    auto size = static_cast<size_t>(std::snprintf(nullptr, 0, format.c_str(), convert_to_c_string(std::forward<Args>(args))...) + 1);
+    if (size <= 0) { throw std::runtime_error("Error during formatting."); }
```

Comment:
Should cast and add +1 after checking for errors. Should not throw for zero as it's a valid size.

Also need to specialize this template function for zero arguments for safety.

Reply:
I'll just wait until your fmt PR merges and then update this one to use the format function you have to handle this correctly.